src/v/utils/external_process.h

```diff
@@ -0,0 +1,262 @@
+     * stderr
+     * @note Will assert if called >1
+     */
+    template<typename Consumer>
+    requires ss::InputStreamConsumer<Consumer, char>
+    void set_stderr_consumer(Consumer consumer) {
+        vassert(!_stderr_consumer.has_value(), "Already set stdout consumer");
+        auto cerr = _process.cerr();
+        _stderr_consumer.emplace(ss::do_with(
+          std::move(cerr),
+          [consumer = std::move(consumer)](
+            ss::input_stream<char>& cerr) mutable {
+              return cerr.consume(std::move(consumer));
+          }));
+    }
+
+    /**
+     * @brief Waits for the process to finish executing
+     *
+     * This will in turn call wait(2) on the child process, awaiting for it
+     * to complete.  This function _must_ be called on all created
+     * `external_process` objects.  This can only be called once.
+     *
+     * @return ss::future<ss::experimental::process::wait_status> The result of
+     * the process
+     * @throw std::system_error Containing an `external_process::error_code`:
+     * * `external_process::process_already_completed` if process already
+     *    completed
+     * * `process_already_being_waited_on` if `wait()` has already been called
+     */
+    ss::future<ss::experimental::process::wait_status> wait();
+
+    /**
+     * @brief Attempts to call SIGTERM and then SIGKILL on the running process
+     *
+     * @note Must call this _after_ calling `wait()`
+     *
+     * @param timeout How long to wait until calling SIGKILL if SIGTERM fails
+     * @return On completion
+     * @throw std::system_error Containing an `external_process::errro_code`
+     * * `external_process::process_already_completed` if process already done
+     * * `external_process::process_already_being_terminated` if `terminate()`
+     *    already called
+     * * `external_process::failed_to_terminate_process` if unable to signal
+     *    process
+     * * `external_process::terminate_called_before_wait` if `terminate()`
+     *    called before `wait()`
+     */
+    ss::future<> terminate(std::chrono::milliseconds timeout);
+
+private:
+    external_process(
+      ss::experimental::process p,
+      std::filesystem::path path,
+      ss::experimental::spawn_parameters params)
+      : _process(std::move(p))
+      , _path(std::move(path))
+      , _params(std::move(params))
+      , _terminate_mutex("external_process::terminate")
+      , _wait_mutex("external_process::wait") {}
+
+private:
+    // The external process
+    ss::experimental::process _process;
+    // Path to the executable that's being executed
+    std::filesystem::path _path;
+    // The parameters for the process
+    ss::experimental::spawn_parameters _params;
+    // Flag indicating that the external process has completed
+    bool _completed{false};
+    // Protects against multiple calls to `terminate()`
+    mutex _terminate_mutex;
+    // Protects against multiple calls to `wait()`
+    mutex _wait_mutex;
```

Comment:
Can we just have a single mutex? Is it important to have two - should you be able to call `terminate` while `wait` has not completed?

Reply:
>should you be able to call `terminate` while `wait` has not completed.

Mike can confirm, but I'm pretty sure that's the only time you call terminate. Assuming I have that right, maybe it should be stated explicitly in the docstring.

Reply:
You have to call `wait()` to ensure that the `wait(2)` system call is issued against the child process (else zombies can happen when the child process dies and doesn't get reaped).  So the idea would be that `wait()` is being serviced and then `terminate()` is called separately.  You must call `terminate()` while `wait()` is being called.

Reply:
> You have to call `wait()` to ensure that the `wait(2)` system call is issued against the child process (else zombies can happen when the child process dies and doesn't get reaped). So the idea would be that `wait()` is being serviced and then `terminate()` is called separately. You must call `terminate()` while `wait()` is being called.

I guess a `semaphore{2}` and assert on the remaining units could work, but it's clear this way.